src/injectors-cache.ts

```diff
@@ -0,0 +1,157 @@
+/**
+ * Contains functions that inject dynamic content.
+ */
+
+import * as brotli from './brotli-wasm-wrapper';
+import {
+  CacheControl,
+  ContentEncoding,
+  HeaderKeys,
+  IncomingCloudflarePropertiesExtended,
+  supportsBrotli,
+} from './headers';
+
+const INJECTORS_CACHE_NAME = 'injectors-cache-v1';
+
+let injectorsCache_: Cache;
+/** */
+async function openInjectorsCache(): Promise<Cache> {
+  if (!injectorsCache_) {
+    injectorsCache_ = await caches.open(INJECTORS_CACHE_NAME);
+  }
+  return injectorsCache_;
+}
+
+/**
+ * Generates a complete cache key for a cache object.
+ *
+ * @param parts - unique list of identifiers for object in cache.
+ * @returns a unique cache key.
+ */
+function cacheKey(...parts: string[]): string {
+  return parts.join(';');
+}
+
+/**
+ * Generates a complete cache key for a Brotli-compressed cache object.
+ *
+ * @param parts - unique list of identifiers for object in cache.
+ * @returns a unique cache key.
+ */
+function cacheKeyBrotli(...parts: string[]): string {
+  return cacheKey(...parts, ContentEncoding.BROTLI);
+}
+
+/**
+ * Gets an already injected Response for the given URL.
+ *
+ * @param url - base file URL that has dynamic content injected into it.
+ * @param key - unique identifiers for this specific injection.
+ * @param cf - incoming Cloudflare properties, used to determine whether to try
+ *     fetching the Brotli pre-compressed version of the URL.
+ * @returns an already injected Response.
+ */
+export async function getCacheFor(
+  url: string,
+  key: string,
+  cf: IncomingCloudflarePropertiesExtended
+): Promise<Response | undefined> {
+  const cache = await openInjectorsCache();
+  return supportsBrotli(cf)
+    ? cache.match(cacheKeyBrotli(url, key))
```

Comment:
If the brotli compression isn't complete, we could try falling back to the plain and serving that.

Reply:
Is it worth building this exception into the code? This falls back into regenerating the dynamic content, which is also a quick action - this would also re-enqueue the cache, so if the plain cache succeeded but the Brotli cache failed, it would retry it this way

Reply:
I think we could use this as the mutex.
- If the brotli item exists, serve it
- If the plain item exists, serve it
- Else, queue brotli compression

If brotli compression fails, we might be able to do a `catch` block to remove the plain item from cache.

Reply:
I don't think we can, my tests show that if the WASM module fails to perform the compression for any reason it immediately halts execution - no catches